Add macro for generating functional components

hamza1311 opened the pull request:
#### Description

This PR adds:
- A macro for generating functional components
- Documents said macro
- Adds placeholder documentation for hooks
- Renames "functional components"
to "function components"

#### Checklist

<!-- For further details, please read CONTRIBUTING.md -->

- [x] I have run `cargo make pr-flow`
  With exception of `test-flow` on `yew-macro`. That requires dependencies to be installed and I didn't change anything from that project anyway.
- [x] I have reviewed my own code
- [x] I have added tests
  <!-- If this is a bug fix, these tests will fail if the bug is present (to stop it from cropping up again) -->
  <!-- If this is a feature, my tests prove that the feature works -->

siku2 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,78 @@
+                let props_type = match inputs.len() {
+                    // If there is one and only one argument passed, use it for props
+                    1 => match inputs.first().unwrap() {
+                        FnArg::Typed(arg) => (*arg.ty).clone(),
+
+                        _ => {
```
This isn't really a big issue but it should be possible to remove the need for this `.clone()` and the above `.unwrap()` by turning `inputs` into an iterator:
```rust
// might need to destructure `func`
let mut inputs = func.sig.inputs.into_iter();

if let Some(first) = inputs.next() {
  // we have a props parameter (which is owned)
} else {
  // use `()` as the properties type
}

for param in inputs {
  // emit an error for each additional input parameter
}
```

siku2 reviewed (changes requested):
As you know there's currently no one in charge of the `yew-functional` project and given that the current features still aren't finished (and documentation is non-existent) I'm hesitant to accept new feature additions. But since it's already here I guess I don't have a choice :upside_down_face:. However, I'm asking you to add documentation for this (only for the macro of course).

> this is something which doesn't need its own tests are already covered by `FunctionProvider` trait. This macro only adds a way to create a struct, implementing the said trait.

This very much needs tests! Compilation tests, that is. We're using [trybuild](https://github.com/dtolnay/trybuild) to test both code that should compile and code that should fail to compile.
[Take a look at how `yew-macro` does it](https://github.com/yewstack/yew/tree/master/yew-macro/tests/props_macro) (the file that actually runs these tests is [here](https://github.com/yewstack/yew/blob/master/yew-macro/tests/props_macro_test.rs)).

You should have pass tests for all the different types of valid inputs (particularly the edge cases) and fail tests for the various error messages that can be generated by your macro.

This should help uncover some unhandled cases and suboptimal error spans.
One case that comes to mind that should fail and currently doesn't is this:
```rust
// the current implementation silently ignores generics
#[functional_component(Comp1)]
fn comp1<P>(props: &P) -> Html where P: Properties { unimplemented!() }

// If we can (and I'm fairly sure it's possible) we should probably support generic functional components.
```
There are other trivial cases like const, async, and extern functions.

Please use the attribute `#![no_implicit_prelude]` on the pass tests to make sure that the macro doesn't use unqualified paths.

hamza1311 commented on examples/README.md:
```diff
@@ -49,6 +49,7 @@ As an example, check out the TodoMVC example here: <https://examples.yew.rs/todo
 | [todomvc](todomvc)           | Implementation of [TodoMVC](http://todomvc.com/)                                                                                   |
 | [two_apps](two_apps)         | Runs two separate Yew apps which can communicate with each other                                                                   |
 | [webgl](webgl)               | Controls a [WebGL canvas](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Tutorial/Getting_started_with_WebGL) from Yew |
+| [counter](counter)           | Simple counter which can be incremented and decremented built using functional components                                          |
```
That's understandable. I'll just remove the example for now.

siku2 replied:
Cool, thanks for understanding

hamza1311 commented on yew-functional/Cargo.toml:
```diff
@@ -18,3 +18,4 @@ yew = { path = "../yew" }
 [dependencies]
 yew = { path = "../yew" }
 wasm-bindgen = "0.2.60"
+yew-functional-macro = { path = "../yew-functional-macro" }
```
Shouldn't this be linted as well?

siku2 replied:
Do you know of a way to lint this?

hamza1311 replied:
I'm not sure, that's part of why I asked. I should've worded that a bit better. Maybe it isn't possible.

siku2 replied:
I don't think the standard tools can do it, but there seems to be a crate for it: [`cargo-sort-ck`](https://crates.io/crates/cargo-sort-ck). I don't think it matters enough to justify using it though.

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,78 @@
+                    }
+                };
+
+                let body = *func.block;
```
I take it out so I don't have to `Box` the field on the struct. It helps read better since you don't have to know why `Block` is `Box`ed in the struct field. I think its better as is, lemme know if this should be changed.

siku2 replied:
I agree it's more readable in the code, but I think the performance of moving something potentially quite large from the heap to the stack isn't ideal. Keeping it boxed should (hopefully) help with compile times, even if it's just a tiny bit.

hamza1311 replied:
Alright, I've `Box`ed it.

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,78 @@
+                    _ => {
+                        return Err(syn::Error::new_spanned(
+                            &func.sig,
+                            "Functional components take only one argument of props struct as value",
```
This is overwritten by forcing of type being a reference.

siku2 replied:
Huh, we still need an error message for >1 parameters, don't we?

hamza1311 replied:
We did. I added a check for that.

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,78 @@
+                        FnArg::Typed(arg) => (*arg.ty).clone(),
+
+                        _ => {
+                            return Err(syn::Error::new_spanned(&func.sig, "invalid argument passed. (hint: first argument must not be a receiver. Functional components take one argument of props struct as value)"));
```
That sounds like a better error message. Changed it.

hamza1311 commented:
I have made the requested changes. Let me know if there's anything I missed. As for tests, I forgot about compilation tests. I will add those soon.

> There are other trivial cases like const, async, and extern functions.

Compiler doesn't let `html!` to be invoked from const functions so I don't think that's an issue. Nothing errors out with neither `async` nor `extern` so there doesn't seem to be a need to have for checks for those.

siku2 commented:
> What exactly needs to be written and where should it go?

The documentation resides in the top level `docs` directory. Let's add a new folder under "concepts" and maybe call it "functional-components". In it, create a markdown file for the derive macro. You can see the other files for reference.
To actually add it to the sidebar, you need to reference the file in [website/sidebars.json](https://github.com/yewstack/yew/blob/master/website/sidebars.json).

See the `website` dir if you want to test the changes locally.

siku2 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,144 @@
+use proc_macro::TokenStream as TokenStream1;
```
don't do this, please. Just use `proc_macro::TokenStream` in the signature, that's much less confusing.
You can just call `.into()` on `quoted` so you don't have to use it there.

hamza1311 replied:
Oh, okay. Fixed that.

siku2 commented:
> Compiler doesn't let `html!` to be invoked from const functions so I don't think that's an issue. Nothing errors out with neither `async` nor `extern` so there doesn't seem to be a need to have for checks for those.

That's exactly the problem. All of these *should* error because the input function doesn't match the trait method. To give an example:
```rust
#[functional_component(Comp)]
async fn comp() -> Html {
  let text = my_async_function().await;
  html! { <span>{ text }</span> }
}
```
The macro won't complain but the compiler sure will because the generated trait implementation is using `await` in a normal method.

siku2 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+
+                            (ty.clone(), ident.clone())
+                        }
+
+                        arg => {
+                            return Err(syn::Error::new_spanned(
+                                arg,
+                                "functional components cannot accept a receiver",
+                            ));
+                        }
```
We might as well make this an exhaustive match if there are only two options.
```suggestion
                        FnArg::Receiver(_) => {
                            return Err(syn::Error::new_spanned(
                                arg,
                                "functional components cannot accept a receiver",
                            ));
                        }
```

siku2 reviewed (commented):
Here's round two. Procedural macros are very tricky and it's easy for errors to slip in so please bear with me.

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+                                ty => return Err(syn::Error::new_spanned(ty, "invalid argument passed. (hint: type must be props struct by reference)"))
+                            };
+
+                            (ty.clone(), ident.clone())
```
That suggestion was when there was only type involved. when there's both type and ident, it can't moved so clone is needed. If you know of a way to not clone type and ident, I'd like to know that.

siku2 replied:
The two values don't overlap so there's no reason why you couldn't move them both out of `arg`. I don't think it makes sense to change this now if we're gonna switch to the transparent `FnArg` approach (where it'll make sense to clone the type to avoid dealing with a self-referential struct) but just for reference:
```rust
// arg is owned because we used `into_iter()`
FnArg::Typed(arg) => {
  // this gives us owned values for `pat` and `ty`
  let PatType { attrs, pat, colon_token: _, ty } = arg;

  // extract owned ident from `pat`
  // extract owned type from `ty`
}
```

hamza1311 replied:
That's good to know.

I switched out to transparent `FnArg` approach but `ty.elem` still needs to be cloned to get the type but I guess that's not a big deal and file as is.

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+                                pat => {
+                                    return Err(syn::Error::new_spanned(
+                                        pat,
+                                        "Cannot obtain ident. This should never happen",
```
That can't happen in this case though. 

```rust
pub fn name(ref props: Props) -> Html;
```
would fail because `Props` must be passed as `&Props`. This is fine as is, I guess?

siku2 replied:
No it's not, because `ref props: &Props` is valid.

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+                    ));
+                }
+
+                let inputs = &func.sig.inputs;
```
Yeah, that sounds like a good idea, I'll change it.

hamza1311 commented:
> That's exactly the problem. All of these _should_ error because the input function doesn't match the trait method. To give an example:
> 
> ```rust
> #[functional_component(Comp)]
> async fn comp() -> Html {
>   let text = my_async_function().await;
>   html! { <span>{ text }</span> }
> }
> ```
> 
> The macro won't complain but the compiler sure will because the generated trait implementation is using `await` in a normal method.
> Even if we didn't care about the suboptimal error message generated by the compiler, we do at least want to test these cases to make sure they behave like we expect them to.

I've made it fail at const, async and extern functions

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,145 @@
+                            let ty = match &*arg.ty {
+                                Type::Reference(ty) => {
+                                    if ty.lifetime.is_some() {
+                                        return Err(syn::Error::new_spanned(&ty.lifetime, "reference must not have life time"));
```
Elided lifetimes and generics are thins that should be left for a later time. These probably won't have much use with these functions anyway.

hamza1311 commented:
Just a thought, this macro doesn't ensure that the return type is `yew::Html`. Is there any way to guarantee that? From what I'm able to find, the most information I can obtain is the `Ident` which doesn't seem to be enough to ensure the return type.

Also, for this documentation mentioned,

> > What exactly needs to be written and where should it go?
> 
> The documentation resides in the top level `docs` directory. Let's add a new folder under "concepts" and maybe call it "functional-components". In it, create a markdown file for the derive macro. You can see the other files for reference.
> To actually add it to the sidebar, you need to reference the file in [website/sidebars.json](https://github.com/yewstack/yew/blob/master/website/sidebars.json).
> 
> See the `website` dir if you want to test the changes locally.

What docs needs to be written? What the macro is and how to use it should be enough, right?

siku2 commented:
> Just a thought, this macro doesn't ensure that the return type is `yew::Html`. Is there any way to guarantee that? From what I'm able to find, the most information I can obtain is the `Ident` which doesn't seem to be enough to ensure the return type.

No, that's not something we can do.
That's something that needs to be tested with compilation tests as well to make sure that the error message is helpful.
There are a few tricks we can use if the error message doesn't make sense.



> What docs needs to be written? What the macro is and how to use it should be enough, right?

Sounds good to me, yeah. The page should ideally also include an example.
You like to bicker about our documentation, now it's your time to deliver :stuck_out_tongue:.

siku2 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,171 @@
+                    ));
+                }
+
+                let ty = match arg {
+                    FnArg::Typed(ref arg) => match &*arg.ty {
```
You can get rid of the `ref` if you want. I don't really care either way but I've heard many users dislike the `ref` pattern:
```suggestion
                let ty = match &arg {
                    FnArg::Typed(arg) => match &*arg.ty {
```
(Note: you have to modify line `101` to use `&arg` as well)

siku2 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,171 @@
+                    let params: TokenStream = inputs.map(|it| it.to_token_stream()).collect();
+                    return Err(syn::Error::new_spanned(
+                        params,
+                        "functional components must take only parameter of props",
```
This error message should be rephrased. Perhaps:
```suggestion
                        "functional components can accept at most one parameter for the props",
```

siku2 reviewed (commented):
Okay, final round. Other than these small things the implementation looks solid!

siku2 commented on yew-functional-macro/tests/functional_macro/no-name-fail.rs:
```diff
@@ -0,0 +1,19 @@
+    a: usize,
+}
+
+#[functional_component()]
```
The error message here should be improved by my suggestion above. Could you also add a test for the following cases:
- invalid Rust ident
- Rust keyword
- a case where there are tokens following the ident: `#[functional_component(a, b, c)]` (I believe this currently doesn't fail and just silently ignores the extra tokens which probably isn't what we want)
- what happens when the component name is the same as the function?

hamza1311 replied:
> (I believe this currently doesn't fail and just silently ignores the extra tokens which probably isn't what we want)

It does fail with `error: unexpected token`. I will add those.

siku2 commented on yew-functional-macro/tests/macro_tests.rs:
```diff
@@ -0,0 +1,7 @@
+#[allow(dead_code)]
```
Could you rename this to something like `functional_component_attr_test` instead? We might add more macros in the future.

siku2 replied:
Did you just rename this back?

siku2 commented on yew-functional-macro/tests/functional_macro/with-props-pass.rs:
```diff
@@ -0,0 +1,21 @@
+#![no_implicit_prelude]
+
+use yew::prelude::*;
+use yew_functional::{functional_component};
+
+#[derive(Clone, Properties, PartialEq)]
+struct Props {
+    a: usize,
+}
+
+#[functional_component(Comp)]
+fn comp(props: &Props) -> Html {
+    html! {
+        <p>
+            { props.a }
+        </p>
+    }
+}
+
+
+fn main() {}
```
The `no_implicit_prelude` tests are most effective when we don't bring anything into scope.
For instance, right now this test wouldn't have caught the unqualified `Html` type the initial version had.

Normally you can just define the test setup in a submodule, but for the `functional_component` attribute that can only really be done for the `Props`.

```suggestion
#![no_implicit_prelude]

#[derive(Clone, PartialEq, ::yew::Properties)]
struct Props {
    a: usize,
}

#[::yew_functional::functional_component(Comp)]
fn comp(props: &Props) -> ::yew::Html {
    ::yew::html! { <p>{ props.a }</p> }
}

fn main() {}
```

siku2 replied:
We could also use a test for the `ref props: &Props` pattern, just to make sure that actually works now.

hamza1311 replied:
I think we should just remove `no_implicit_prelude`. Changing it to how you suggested makes the code too verbose. Or is there any reason to keep this?

siku2 replied:
The reason for this is to test the hygiene of the macro. It's very easy for procedural macros to accidentally use items from the local scope (the original version of this PR had one such error, for example). Needless to say that that isn't a good idea; dealing with the resulting errors can be very confusing for the user and annoying to fix for the developers so I would like to ensure we're guarded against this going forward.

Yes, it makes the passing tests a bit more verbose, but this is an integral part of the test in the first place.
It only needs to be done for the passing tests so it really shouldn't be that bad.

hamza1311 replied:
I've changed the two passing tests to use it.

siku2 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,175 @@
+                format!("invalid name for component provided ({})", e.to_string()),
+            )
+        })?;
+        let function_name = format_ident!("Function{}", component_name);
```
In your opinion, is this `FunctionXYZ` struct something that the user should be allowed to access (i.e. something that's part of the "official API")?

hamza1311 replied:
I don't think we (will) have much of a choice here. Consider this code:
```rust
struct XYZ;

pub type ABC = XYZ;
```
This results in warning:
```
warning: private type `XYZ` in public interface (error E0446)
 --> src/main.rs:3:1
  |
3 | pub type ABC = XYZ;
  | ^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(private_in_public)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #34537 <https://github.com/rust-lang/rust/issues/34537>
```

This already is a warning and will become an error so I don't think we should make a part of "unofficial" or "private" API.

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8fec43abbb91330f6ecadb23949b3ad0)

siku2 replied:
That's not what I was asking though. This isn't a question of whether we can make the struct itself private but rather whether we "define" this behaviour (i.e. document it and tell the user about it). The alternative is that we treat it as an implementation detail by giving the ident a "mixed site" span (which should make it so the user code can't access it), moving it into a `mod __impl` to discourage the user from using it, or just giving it some obscure name.

hamza1311 replied:
That trait does not provide any functionality useful to the future. Unless that changes, I don't think there's a reason to expose it. The only downside I can think of is, when IDE support for proc_macros is good, it'll be a useless item in the autocomplete suggestions list.

How about we make it private right now and change it if there's a need for it?

siku2 replied:
Yeah, that's what I had in mind too. Sadly "def site" spans are still nightly only. By giving the function provider ident `Span::def_site()` it would've been completely inaccessible from user code but oh well...

I do have a pretty good idea though! Rust allows us to declare a struct and a function with the same name (because a function is never a type and a struct isn't a function so they never overlap). So let's just give the `FunctionProvider` struct the same ident as the function has. The benefit of doing it this way is that we don't need to worry about name conflicts (if the user already has a `FunctionXYZ` type in scope the current approach errors), the identifier is essentially invisible, and if we do ever decide to expose it as part of the public API we already have a well defined and obvious naming scheme.

This would look like this: 
```rust
#[doc(hidden)]
#[allow(non_camel_case_types)]
#vis struct #fn_ident;
```
where `fn_ident` is taken from `sig.ident`.

The only thing this changes is the error message for the case where the component name is the same as the function name which becomes: "the name `xyz` is defined multiple times". We can easily handle this case by emitting a custom error for this though.

hamza1311 replied:
I'd prefer just giving it an obscure name or moving it to a private module with `#[doc(hidden)]`. Using the function name to hide it just feels wrong.

siku2 replied:
I'm sure you already know that your reasoning here isn't valid :smile:. The problem with the other approaches is that error messages (or any message that contains the type) is gonna look bad and even worse, confusing.
By giving it the same ident as the function itself it's trivial to tell which function we're talking about.

hamza1311 replied:
I've made it use the function name but the error message is suboptimal when the function name and component are same. How do you think that should be handled?

siku2 replied:
You can just compare `name == component_name` and if they are the same, emit an error pointing at the component name.
The problem you might run into is how to actually emit an error from there.

I would move the code generation code to a separate function which returns `syn::Result<TokenStream>` and then you can call `.to_compile_error()` on the `syn::Error` to turn it into a `TokenStream` (proc macro errors are literally just invocations of the `compile_error!` macro):
```rust
fn implement_functional_component(name: FunctionalComponentName, component: FunctionalComponent) -> syn::Result<TokenStream> {
    let FunctionalComponentName { component_name } = name;
    let FunctionalComponent {
        block,
        props_type,
        arg,
        vis,
        attrs,
        name,
        return_type,
    } = component;

    if component_name == name {
        return Err(syn::Error::new_spanned(component_name, "component must not have the same name as the function"));
    }
    
    let ret_type = quote_spanned!(return_type.span() => ::yew::html::Html);

    quote! {
        #[doc(hidden)]
        #[allow(non_camel_case_types)]
        #vis struct #name;

        impl ::yew_functional::FunctionProvider for #name {
            type TProps = #props_type;

            fn run(#arg) -> #ret_type {
                #block
            }
        }

        #(#attrs)*
        #vis type #component_name = ::yew_functional::FunctionComponent<#name>;
    }
}

#[proc_macro_attribute]
pub fn functional_component(
    attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let name = parse_macro_input!(attr as FunctionalComponentName);
    let component = parse_macro_input!(item as FunctionalComponent);
    
    implement_functional_component(name, component)
        .unwrap_or_else(|err| err.to_compile_error())
        .into()
}
```

siku2 reviewed (commented):
Thanks for adding the tests, they're great! I appreciate how you split them up into different files, that's much easier to work with than when everything's in a single file.

If it's not too much to ask, could you run the test files through Rustfmt?
This doesn't happen automatically, sadly, and some of them could use a bit of formatting :P

I've added a few comments for test cases that I would like to include as well.
There are also two error messages that aren't tested yet:
- function that has a receiver (self). This might need to be in an `impl` block for Rust to even accept it.
- attribute being applied to something other than a function (only one case like a struct should totally suffice)

hamza1311 commented on yew-functional-macro/tests/functional_macro/bad-return-type-fail.stderr:
```diff
@@ -0,0 +1,12 @@
+error[E0308]: mismatched types
+  --> $DIR/bad-return-type-fail.rs:10:24
+   |
+10 |   fn comp(props: &Props) {
```
It's the same error, just with different type.  
```rust
#[functional_component(Comp)]
fn comp(props: &Props) -> u32 {
    let _x = html! {
        <p>
            { props.a }
        </p>
    };
    1
}
```
`stderr` file:
```
error[E0308]: mismatched types
  --> $DIR/bad-return-type-fail.rs:16:5
   |
9  | #[functional_component(Comp)]
   | ----------------------------- expected `yew::virtual_dom::vnode::VNode` because of return type
...
16 |     1
   |     ^ expected enum `yew::virtual_dom::vnode::VNode`, found integer
```

hamza1311 commented:
> I appreciate how you split them up into different files, that's much easier to work with than when everything's in a single file.

With the last commit, I've added multiple tests in one file where appropriate .

Also, I added some documentation so let me know if there's anything that needs to me changed.

siku2 commented on docs/concepts/functional-components.md:
```diff
@@ -0,0 +1,21 @@
+description: Introduction to functional components 
+---
+
+:::note
+Functional components are currently unreleased.
+:::
```
The previous section had a link to the project board, let's add that here too.
```suggestion
:::warning
Functional components are still under development.
Take a look at the [project board](https://github.com/yewstack/yew/projects/3) for details and consider helping out.
:::
```

siku2 replied:
Why did you place the link further down? The project board isn't specific to hooks or their documentation, it's about yew-functional as a whole.
My suggestion also changes the admonition from "note" to "warning" because we really don't want users to use them yet.

hamza1311 replied:
It was me trying to add some docs/placeholders for hooks docs too. I've moved the link back above now.

siku2 commented on docs/concepts/functional-components/macro.md:
```diff
@@ -0,0 +1,33 @@
+---
+title: #[functional_component]
+description: The #[functional_component] macro
```
I think we should call it "attribute" instead of "macro".
"macro", to me, makes it sound like it would be used like this:
```rust
functional_component! {
  fn my_component() -> Html {
    html! {}
  }
}

siku2 commented on docs/concepts/functional-components/macro.md:
````diff
@@ -0,0 +1,33 @@
+### Example
+
+```rust
+#[derive(Clone, Properties, PartialEq)]
+struct Props {
+    a: usize,
+}
+
+#[functional_component(Comp)]
+fn comp(props: &Props) -> Html {
+    html! {
+        <p>
+            { props.text }
+        </p>
+    }
+}
````
How about we use a slightly simplified version of the original "counter" example you had here?

hamza1311 replied:
The counter example would require the use of hooks. Is that something that should be used in documentation without any mentioning of what they are/do?

siku2 replied:
Of course. It's not ideal that hooks aren't documented yet, of course, but all of this isn't supposed to be used yet anyway (it might even be merged into the yew crate before being released).

hamza1311 replied:
Okay, I might add *some* documentation with PR. This changed example will be a little big though.

siku2 replied:
Huh, I didn't want to give the impression that you had to add documentation for hooks. If you want to then sure go ahead (though maybe consider doing it in a separate PR to keep this focused).

> This changed example will be a little big though.

That's why I said slightly simplified. But the original example wasn't that long so it shouldn't be that much of a problem.

hamza1311 replied:
I have already added some placeholders for those. The documentation for those is something to be done on a later day though.

siku2 commented on yew-functional-macro/tests/functional_macro/bad-return-type-fail.rs:
```diff
@@ -0,0 +1,28 @@
+}
+
+#[functional_component(Comp)]
+fn comp(props: &Props) -> u32 {
+    let _x = html! {
+        <p>
+            { props.a }
+        </p>
+    };
```
This isn't needed for the test
```suggestion
fn comp(_props: &Props) -> u32 {
```

siku2 commented on yew-functional-macro/tests/functional_macro/bad-return-type-fail.stderr:
```diff
@@ -0,0 +1,21 @@
+16 | | }
+   | |_^ expected enum `yew::virtual_dom::vnode::VNode`, found `()`
+
+error[E0308]: mismatched types
+  --> $DIR/bad-return-type-fail.rs:25:5
+   |
+18 | #[functional_component(Comp)]
+   | ----------------------------- expected `yew::virtual_dom::vnode::VNode` because of return type
+...
+25 |     1
+   |     ^ expected enum `yew::virtual_dom::vnode::VNode`, found integer
```
Not perfectly happy with this error message. Should we do something about it?

hamza1311 replied:
What can be done about it, considering that checking return type is impossible?

siku2 replied:
Well the issue here is that the span points at the macro invocation (i.e. call site). If we give the return type of the trait method the same span as the input function return type it will point there instead.

hamza1311 replied:
How would that be done?

siku2 replied:
First, extract the return type from the signature. As an additional benefit we can now also provide a custom error message for when the input function doesn't have a return type at all.
```rust
let return_ty = match sig.output {
    ReturnType::Default => {
        return Err(syn::Error::new_spanned(sig, "function must return `Html`"));
    }
    ReturnType::Type(_, ty) => ty,
};
```

And in the output we apply the span from the input return type to our `Html`:
```rust
let return_ty = quote_spanned!(return_ty.span()=> ::yew::html::Html);

let quoted = quote! {
    #vis struct #function_name;

    impl ::yew_functional::FunctionProvider for #function_name {
        type TProps = #props_type;

        fn run(#arg) -> #return_ty {
            #body
        }
    }

    #vis type #component_name = ::yew_functional::FunctionComponent<#function_name>;
};
```

The new error will look like this:
```rust
error[E0308]: mismatched types
  --> $DIR/bad-return-type-fail.rs:25:5
   |
19 | fn comp(props: &Props) -> u32 {
   |                           --- expected `yew::virtual_dom::vnode::VNode` because of return type
...
25 |     1
   |     ^ expected enum `yew::virtual_dom::vnode::VNode`, found integer
```
That's pretty nice, isn't it?

hamza1311 replied:
That is indeed nice. Fixed that.

siku2 replied:
You ignored the first part of my suggestion. The error spans are significantly worse because of that.

siku2 replied:
You missed another thing for the `ReturnType::Default` case. You're using the span from `sig.output` but that doesn't have a span because there is no output type, so it falls back to call site. If you look at my suggestion again, you'll see that I'm using `sig` for the span. That way the error points at the `fn comp()` part which is probably the best we can do.

hamza1311 replied:
Oops, proc macros really are fragile.

siku2 commented on yew-functional-macro/tests/functional_macro/multiple-param-fail.rs:
```diff
@@ -0,0 +1,20 @@
+}
+
+#[functional_component(Comp)]
+fn comp(props: &Props, invalid: String) -> Html {
```
Did I miss something or did you forget to add this?

hamza1311 replied:
I think I missed that. I will just add it.

siku2 commented on docs/concepts/functional-components.md:
```diff
@@ -0,0 +1,21 @@
+
+## Creating functional components
+
+Functional components are created using the [`#[functional_component]`](functional-components/macro.md) attribute macro.
```
All the documentation you've written so far (including the docstring) doesn't really tell the user what's happening under the hood. Could you please add a few sentences explaining the `FunctionProvider` trait and the `FunctionComponent` struct just to give some context. There are cases where you'd have to implement it manually and giving the user a general idea of what an attribute actually does never hurts.

hamza1311 replied:
What are those cases?

I think doing this is similar to documenting how `html!` uses `VNode`. If this is something important, it can be added to the docstring. I'm not a fan of adding implemention details to website though.

siku2 replied:
`FunctionProvider` isn't not just an implementation detail. It's part of the public API surface (for better or worse).
By giving an overview of how things work it's easier to explain why certain restrictions exist and what to watch out for. It's also valuable for potential contributors because they already have an idea of how the magic works.


> I think doing this is similar to documenting how `html!` uses `VNode`.

That is something we're planning to do (and already started doing at some point). The problem there is that Yew's `virtual_dom` is nowhere near as stable yet and a lot of users don't fully understand it (which is obviously due to the lack of documentation) so there aren't that many people even capable of writing it.

siku2 reviewed (commented):
Reading the documentation made me realize that we have a bit of a naming conflict going on between "function*al* components" and "function components". I personally don't really care either way (though React calls them "function components" and to me that makes more sense grammatically) but I feel like we should make this consistent. I assume you prefer "functional" given that that's what you went with?

---

Also, I realised that we forgot about the transfer of documentation:
```rust
/// My documented component
#[functional_component(Foo)]
fn comp() -> Html {
  html! {}
}
```

The documentation for the function should be applied to the `FunctionComponent` type, otherwise component libraries won't be able to use this.
To do this I guess we'll have to transfer the attributes over to the type declaration. ~~We can just select the `#[doc ...]` attributes so we don't move over the `#[functional_component]` one as well.~~

siku2 commented:
Btw, could you copy  `Makefile.toml` from `yew-macro` over to `yew-functional-macro`? You can literally just keep it 1:1.

hamza1311 commented:
For passing docs, I'll just pass all the `attrs` as is to the `type` generated so if there's any other attributes that needs to passed, those can be passed too.

> Btw, could you copy `Makefile.toml` from `yew-macro` over to `yew-functional-macro`? You can literally just keep it 1:1.

Sure!

siku2 commented:
> I'll just pass all the `attrs` as is to the `type` generated

Doesn't that also copy the `#[functional_component]` attribute?

siku2 commented:
The question of whether we adopt "function component" (currently used by `yew-functional` and also React)  or "functional component" still stands. I'd love to hear your thoughts on this. Right now I'm much in favour of "function component" but I'm also fine with "functional" as long as we use it consistently.

hamza1311 commented:
> The question of whether we adopt "function component" (used by `yew-functional` and also React)  or "functional component" still stands. I'd love to hear your thoughts on this. Right now I'm much in favour of "function" but I'm also fine with "functional" as long as we use it consistently.

I think we should go for "function*al* components" since it seems more grammatically correct, already used at almost all places in code and documention, and familiar to anyone coming from React. I'm in the favor of using that.

Why do you prefer "function components"?

siku2 commented:
- Yew already has a strong precedent for "function". The backing implementation function components is called `FunctionProvider` and `FunctionComponent`. Yewtil's component macro is also called `function_component`. 
- React officially uses "function" in both code and documentation. I honestly have no idea why "functional" is so prevalent in the community
- this isn't a great reason, but "function" is shorter to type :P
- after looking at several definitions for "functional" I didn't find any that actually makes sense in our context. "functional" doesn't seem to mean that something is function-like (or anything to do with our kind of functions). "function component" on the other hand perfectly describes that it's a component based on a function.

hamza1311 commented:
I guess we should just replace "functional" with "function" then. Should that be changed now or later?

Also, is [this](https://yew.rs/docs/en/next/advanced-topics/optimizations#pure-components) still relevant now or should that be removed?

siku2 commented:
Let's do it now, it should be fairly straightforward and I think after that the implementation is finally done and we can focus on the documentation. 

Pure components are still a thing and they're not related to function components, let's leave it for now. If you're talking about the following section, I think we can remove that

hamza1311 commented:
Replaced  completely replace "functional" with `function` with latest commit *(I just realized I got that the other way around in my commit message... oops)*

I set the crate name to `yew-function-components` because `yew-function` doesn't really make any sense. Is it good now?

siku2 commented:
> I set the crate name to `yew-function-components` because `yew-function` doesn't really make any sense.

Completely agree that `yew-function` is out of the question, but let's just keep it as `yew-functional` for now. My plan is to merge all of this into the `yew` crate anyway and even if that doesn't happen, we can always rename it later on.
`yew-function-components` is way too long and creates a messy diff.

siku2 reviewed (commented):
Apart from the return type span stuff these should hopefully be the last comments for the code.

After that we can finally move on to the documentation :tada:

hamza1311 commented on yew-function-components-macro/src/lib.rs:
```diff
@@ -0,0 +1,215 @@
+                } = func;
+
+                if !sig.generics.params.is_empty() {
+                    // TODO maybe find a way to handle those
```
Can we even fully support generics? I'm not sure if `FunctionProvider::run` would accept them as is.

siku2 replied:
Yes, the following input:
```rust
fn foo<P>(_props: &P) -> Html where P: Properties + PartialEq {
    todo!()
}
```

Can be turned into this:
```rust
// generate a tuple struct with a phantom data field for each generic parameter
struct Foo<P>(::std::marker::PhantomData<P>);

// we only need the `where` guard on the trait implementation
impl<P> FunctionProvider for Foo<P> where P: PartialEq + Properties {
    type TProps = P;

    fn run(_props: &P) -> Html {
        todo!()
    }
}

// because this is a type alias we don't need the guards here
type FooComponent<P> = FunctionComponent<Foo<P>>;
```

This can then be used like this you'd expect:
```rust
html! { <FooComponent<()> /> }
```

siku2 replied:
The only problem is that because Rust allows us to use tuple structs with a function-like syntax we can no longer give the struct the same name as the function. I didn't consider this back when I suggested this...

hamza1311 replied:
I think we should leave this for later. I can't think of a situation where generics wouldn't be needed for function components. What do you think?

siku2 replied:
I think we should do that in a follow up PR though. The code review here is pretty much finished and taking it for another spin would just cause more work.

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,211 @@
+    if function_name == component_name {
+        return Err(syn::Error::new_spanned(
+            component_name,
+            "The function name and component name must not be the same",
```
> Right, thanks for ignoring my previous suggestion yet again.

I'm sorry, I didn't mean to. I really didn't notice that part.

siku2 replied:
It turned out a lot snarkier than it was intended to be, sorry :smile:

hamza1311 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,211 @@
+                    ));
+                }
+
+                let mut inputs = sig.inputs.clone().into_iter();
```
oops, yeah, I confused that block with the one above it. It should've just gone up in the first place.

siku2 commented on docs/concepts/function-components.md:
```diff
@@ -0,0 +1,39 @@
+
+## Creating function components
+
+function components are created using the [`#[function_component]`](function-components/macro.md) attribute.
```
```suggestion
The easiest way to create a function component is to add the [`#[function_component]`](function-components/attribute.md) attribute to a function.

// TODO: let's insert a really small example here (literally just like `html! { "Hello World" }` or something) just to give the reader a taste of what they look like

### Under the hood

function components consists of two parts.
First, the `FunctionProvider` trait which is comparable to the `Component` trait but it only has a single method called `run`.
The second part is the `FunctionComponent` struct which wraps around the `FunctionProvider` type and turns it into an actual `Component`. 
The `#[function_component]` attribute essentially just implements `FunctionProvider` for you and exposes it wrapped in `FunctionComponent`.
```

siku2 commented on yew-functional/src/lib.rs:
```diff
@@ -7,6 +7,37 @@ use yew::{Component, ComponentLink, Html, Properties};
 
 mod use_context_hook;
 pub use use_context_hook::*;
+/// This macro autogenerates the requirements for function components which allows any function to
+/// be annotated and used as a Yew component.
+///
+/// Functions to which this macro is applied to **must** return `Html` and can optionally take an
+/// argument for props. The name for the component is passed as an attribute to the macro.
+///
+/// This attribute generates a struct which implements [`FunctionProvider`] trait.
+/// A type alias `FunctionComponent<CreatedStruct>` is then created.
+/// [`FunctionComponent`] is a struct which implements `yew::Component`
+/// and handles the magic required for making function components work.
+/// In most cases, you don't need to use the trait implementation,
+/// all you need is the aliased `type` which can be used as any other component.
```
```suggestion
/// This attribute creates a function component from a normal Rust function.
///
/// Functions with this attribute **must** return `Html` and can optionally take an argument for props.
/// Note that the function only receives a reference to the props.
/// 
/// When using this attribute you need to provide a name for the component: `#[function_component(ComponentName)]`.
/// The attribute will then automatically create a [`FunctionComponent`] with the given identifier which you can use like a normal component.
```

siku2 reviewed (commented):
Alright, I think the code is finally finished :partying_face:
Now we just gotta work out some of the kinks in the documentation and then we're good to go!

I left a ~~few~~ lot of suggestions to expand the documentation a bit. Documentation is of course highly subjective so feel free to rewrite as you see fit.

There's also a critical issue with the pass tests which needs to be fixed!

hamza1311 commented on docs/concepts/function-components/pre-defined-hooks.md:
```diff
@@ -0,0 +1,27 @@
+description: The pre-defined hooks that Yew comes with 
+---
+
+## `use_state`
+
+
+## `use_ref`
+
+
+## `use_reducer`
+
+
+## `use_reducer_with_init`
+
+
+## `use_effect`
+
+
+## `use_effect_with_deps`
```
I'm thinking of adding these docs after this PR. Probably adopting some of react documentation.

siku2 replied:
Ah right, if you can commit to adding this afterwards that's even better. I just don't want to end up with this empty scaffolding for too long :)

hamza1311 replied:
I'll work on that soon after this is finished.

hamza1311 commented on yew-functional-macro/tests/function_attr_test.rs:
```diff
@@ -0,0 +1,7 @@
+fn tests() {
+    let t = trybuild::TestCases::new();
+    t.pass("tests/function_macro/*pass");
+    t.compile_fail("tests/function_macro/*fail.rs");
```
Oh well, I was also surprised why these tests didn't fail. It turns out they were never run.

hamza1311 commented on docs/concepts/function-components.md:
```diff
@@ -0,0 +1,39 @@
+---
```
Sure

hamza1311 commented:
The project name is still "Functional components" which should probably also updated to "Function components"

siku2 commented:
> The project name is still "Functional components" which should probably also updated to "Function components"

Good point, renamed it.

siku2 commented on docs/concepts/function-components.md:
````diff
@@ -0,0 +1,57 @@
+```rust
+#[function_component(HelloWorld)]
+fn hello_world() -> Html {
+    html! {
+        { "Hello world" }
+    }
````
We can simplify this down to this:
```suggestion
    html! { "Hello world" }
```

siku2 replied:
It's fine to leave it as is if you prefer it that way, but I was thinking it could look something like this:
```rust
## Creating function components

The easiest way to create a function component is to add the [`#[function_component]`](function-components/attribute.md) attribute to a function.

```rust
#[function_component(HelloWorld)]
fn hello_world() -> Html {
    html! { "Hello world" }
}
<CODE BLOCK END cuz GitHub bad>

```

I don't really think of this as an "example", more of a "visual aid" to give some context to the previous sentence. It makes it clear what we mean by "attribute" (some users might not be familiar with them), and already establishes some familiarity with the required signature.

hamza1311 replied:
That is a valid point, especially since there's an example on the other page

siku2 commented on yew-functional-macro/src/lib.rs:
```diff
@@ -0,0 +1,211 @@
+                            if ty.lifetime.is_some() {
+                                return Err(syn::Error::new_spanned(
+                                    &ty.lifetime,
+                                    "reference must not have life time",
```
Noticed a typo while reading through the error messages:
```suggestion
                                    "reference must not have a lifetime",
```

siku2 commented on docs/concepts/function-components.md:
```diff
@@ -0,0 +1,57 @@
+On its own that would be quite limiting because you can only create pure components, but that's where hooks come in.
+Hooks allow function components to use state and other Yew features without implementing the `Component` trait.
+
+## Example:
```
We don't use punctuation on titles (the only exception is a question mark)
```suggestion
## Example
```

siku2 commented on docs/concepts/function-components.md:
```diff
@@ -0,0 +1,57 @@
+
+### Under the hood
+
+function components consists of two parts.
```
```suggestion
Function components consists of two parts.
```
(Unless this was deliberate, but in that case we would have to change it in some other places)

siku2 commented on docs/concepts/function-components/attribute.md:
```diff
@@ -0,0 +1,57 @@
+
+
+The `#[function_component(_)]` turns a normal Rust function into a function component.
+Functions with the attribute have to return  `Html` and may take a single parameter parameter for the type of props the component should accept.
```
Ah yes, the famous "parameter parameter"
```suggestion
Functions with the attribute have to return  `Html` and may take a single parameter for the type of props the component should accept.
```

siku2 reviewed (commented):
I think this is it!

Oh and just because I don't want to scroll up to find the preview link every time: <https://yew-rs--pr1638-functional-derive-hjo24zn4.web.app/docs/en/next>

siku2 commented on docs/concepts/function-components/attribute.md:
```diff
@@ -41,7 +41,9 @@ pub fn rendered_at(props: &RenderedAtProps) -> Html {
     let (counter, set_counter) = use_state(|| 0);
-    let onclick = Callback::from(move |_| set_counter(*counter + 1));
+
+    let counter1 = counter.clone();
+    let onclick = Callback::from(move |_| set_counter(*counter1 + 1));
```
Ah right, should've realised that. Thanks for checking!

One little trick I like to use that makes this a bit prettier is this:
```rust
let onclick = {
  // also note the use of `Rc::clone` which is how Rust recommends you clone `Rc`s
  let counter = Rc::clone(counter);
  Callback::from(move |_| set_counter(*counter + 1))
};
```

It's slightly more verbose and might require some getting used to but this has a few benefits:
- no need to come up with a new variable name every time
- It's immediately clear that the cloning and callback are related
- the cloned variable's scope is limited to the scope so you can't accidentally use it somewhere else

hamza1311 replied:
Huh, is there a reason why `Rc::clone(&counter)` is better than `counter.clone()`?

siku2 replied:
Because `.clone` doesn't tell you much about what's actually happening. Cloning an `Rc` is comparably cheap and is linear in complexity, something that isn't true for strings or vectors. It usually makes sense to optimize away cloning as much as possible but that isn't necessary for `Rc`s and by using `Rc::clone(_)` we make it explicit that that's what we're doing.

And also, from the official docs:
> The Rc::clone(&from) syntax is the most idiomatic because it conveys more explicitly the meaning of the code. In the example above, this syntax makes it easier to see that this code is creating a new reference rather than copying the whole content of foo.

hamza1311 replied:
That's good to know

siku2 reviewed (approved):
Alright, I think we're done!
I'm just gonna wait until tomorrow to merge this in case something important comes up or someone else wants to take a look at the changes
